sockeye/data_io.py

```diff
@@ -401,29 +420,44 @@ def __init__(self,
         self.data_source = [[] for _ in self.buckets]
         self.data_target = [[] for _ in self.buckets]
         self.data_label = [[] for _ in self.buckets]
+        self.data_label_averge_len = [0 for _ in self.buckets]
+
+        # Per-bucket batch sizes [num seq, num word]
+        # If not None, populated as part of assigning to buckets
+        self.bucket_batch_sizes = bucket_batch_sizes
+        # Batch size that pairs with default bucket shape to accommodate any bucket's batch size
+        self.default_bucket_batch_size = 0
 
         # assign sentence pairs to buckets
         self._assign_to_buckets(source_sentences, target_sentences)
 
         # convert to single numpy array for each bucket
         self._convert_to_array()
 
+        # "Staging area" that needs to fit any size batch we're using by total number of elements
         self.provide_data = [
-            mx.io.DataDesc(name=source_data_name, shape=(batch_size, self.default_bucket_key[0]), layout=C.BATCH_MAJOR),
-            mx.io.DataDesc(name=target_data_name, shape=(batch_size, self.default_bucket_key[1]), layout=C.BATCH_MAJOR)]
+            mx.io.DataDesc(name=source_data_name,
+                           shape=(self.default_bucket_batch_size, self.default_bucket_key[0]),
+                           layout=C.BATCH_MAJOR),
+            mx.io.DataDesc(name=target_data_name,
+                           shape=(self.default_bucket_batch_size, self.default_bucket_key[1]),
+                           layout=C.BATCH_MAJOR)]
         self.provide_label = [
-            mx.io.DataDesc(name=label_name, shape=(self.batch_size, self.default_bucket_key[1]), layout=C.BATCH_MAJOR)]
+            mx.io.DataDesc(name=label_name,
+                           shape=(self.default_bucket_batch_size, self.default_bucket_key[1]),
+                           layout=C.BATCH_MAJOR)]
```

Comment:
Yes the metrics will report the same because we collapse batch and time dimension in the softmax output.
I wasn't aware of MXNet's behavior with different batch sizes for batches passed to a module. I think I need to better understand this a bit to be able to completely follow this PR :)

Reply:
I couldn't find concrete documentation for this anywhere, but the working idea is that:
1. Each DataBatch includes a DataDesc with its own provide_data shape.
2. That means in principle MXNet should be able to handle batches of different sizes correctly.
3. We need a single provide_data shape for the DataIter so that memory can be reserved for staging batches.
4. If we make this area large enough to fit the largest batch, MXNet appears to do the right thing.

Reply:
Sounds good! This is in line with what we already to with bucketing:
dataIter.provide_data should use the default/largest bucket key so that parameter and memory sharing for different graphs is working properly. Every batch however, has its own bucket_key member that determines which graph the batch should be routed to.

In your case, I would assume that MXNet requires to know the maximum (batch) size once for each graph, allocates the respective memory, and then dynamically reshapes the graph based on the batch size of each batch, reusing already allocated memory.

Regarding bucketing, I think @tdomhan did an investigation once that if one doesn't use the right default bucket key when creating the module, MXNet does not complain but just silently does the wrong thing :)
We should make sure nothing like this can happen here, either by some empirical evidence that this is doing the right thing or by looking at the module code and try to understand the logic. On a related note: does this affect speed at all?

Reply:
memory sharing isn't well documented in mxnet. The link only is part of how memory sharing is implemented. Essentially this is for all ndarrays that are in python space. Additionally there is logic for sharing ndarrays allocated in C++ land between the C++ executor objects. The executor of the default bucket key is where ndarrays for all other executors come from. The logic is sort of like this pseudo code:
```
unused_ndarrays = default_executor.ndarrays
for ndarray_size in ndarrays_needed_for_executor:
  ndarray = unused_ndarrays.pop_smallest_that_fits(ndarray_size)
  if ndarray == None:
     ndarray = new NDarray()

```
This is happening somewhere here:
https://github.com/apache/incubator-mxnet/blob/master/src/executor/graph_executor.cc#L917 


Bottom line is that the ndarrays of the executor for the default bucket key need to be large enough to fit ndarrays of all other executors. If so no additional memory is required.

Reply:
In any case this is something that has to be tried out. MXNet will only complain if the batch size of the default bucket is too small. However, if it can't reuse NDArrays internally in the graph executor there won't be any warnings, but rather it will just silently allocate new arrays. So in practice running sockeye with and without the `--no-bucketing` arguments is the easiest way of telling if the memory sharing is working correctly (memory consumption should not change much). A safe option might be to use the largest batch size of any of the buckets, but that might use too much memory.